tirocks/src/db/imp.rs

```diff
@@ -0,0 +1,726 @@
+            ffi_call!(crocksdb_approximate_sizes_cf(
+                self.get_ptr(),
+                opt.as_ptr(),
+                cf.get_ptr(),
+                raw_ranges.as_ptr(),
+                raw_ranges.len() as i32,
+                sizes.as_mut_ptr(),
+            ))?;
+            sizes.set_len(raw_ranges.len());
+            Ok(sizes)
+        }
+    }
+
+    /// The method is similar to [`approximate_sizes`], except it returns approximate number
+    /// and size of records in memtables.
+    pub fn approximate_mem_table_stats(
+        &self,
+        cf: &RawCfHandle,
+        start_key: &[u8],
+        end_key: &[u8],
+    ) -> (u64, u64) {
+        unsafe {
+            let raw_range = range_to_rocks(&start_key, &end_key);
+            let (mut count, mut size) = (0, 0);
+            tirocks_sys::crocksdb_approximate_memtable_stats_cf(
+                self.get_ptr(),
+                cf.get_ptr(),
+                &raw_range,
+                &mut count,
+                &mut size,
+            );
+            (count, size)
+        }
+    }
+
+    /// Destroy the contents of the specified database. Be very careful using this method.
+    #[inline]
+    pub fn destroy(
+        path: impl AsRef<Path>,
+        options: &RawOptions,
+        cfs: &[(impl AsRef<str>, impl AsRef<RawCfOptions>)],
+    ) -> Result<()> {
+        unsafe {
+            let mut cf_names = Vec::with_capacity(cfs.len());
+            let mut cf_opts = Vec::with_capacity(cfs.len());
+            for (name, opt) in cfs {
+                cf_names.push(r(name.as_ref().as_bytes()));
+                cf_opts.push(opt.as_ref().as_ptr());
+            }
+            ffi_call!(crocksdb_destroy_db(
+                path.path_to_slice(),
+                options.as_ptr(),
+                cf_names.as_ptr(),
+                cf_opts.as_ptr(),
+                cfs.len(),
+            ))
+        }
+    }
+
+    #[inline]
+    pub fn repair(
+        path: impl AsRef<Path>,
+        options: &RawDbOptions,
+        cfs: &[(impl AsRef<str>, impl AsRef<RawCfOptions>)],
+    ) -> Result<()> {
+        unsafe {
+            let mut cf_names = Vec::with_capacity(cfs.len());
+            let mut cf_opts = Vec::with_capacity(cfs.len());
+            for (name, opt) in cfs {
+                cf_names.push(r(name.as_ref().as_bytes()));
+                cf_opts.push(opt.as_ref().as_ptr());
+            }
+            ffi_call!(crocksdb_repair_db(
+                path.path_to_slice(),
+                options.as_ptr(),
+                cf_names.as_ptr(),
+                cf_opts.as_ptr(),
+                cfs.len(),
+            ))
+        }
+    }
+}
+
+unsafe impl Sync for RawDb {}
+unsafe impl Send for RawDb {}
+
+/// The safe wrapper of `rocksdb::DB` that manages the required lifetime of resources.
+#[derive(Debug)]
+pub struct Db {
+    ptr: *mut rocksdb_DB,
+    _env: Option<Arc<Env>>,
+    comparator: Vec<Arc<SysComparator>>,
```

Comment:
You should probably just keep a copy of options. There're other objects that supplied to RocksDB as raw pointers, it's not easy to keep track of them.

Reply:
That's not true. wal_filter and compaction_filter are raw pointers. My point is it's hard to keep track of these manually.

Reply:
`wal_filter` is not supported yet. `compaction_filter` is not allowed to be set via tirocks, instead, compaction filter factory should be used, which is a shared_ptr.

> My point is it's hard to keep track of these manually.

Actually, it should be easy. Because every time adding support to a new field, we have to write all the codes manually. So it's impossible to miss any of these.

Reply:
Fair.